packages/assets-controllers/src/TokenRatesController.ts

```diff
@@ -421,194 +367,195 @@ export class TokenRatesController extends StaticIntervalPollingController<TokenR
-   * rates in a known currency instead, then converts those rates using the
-   * exchange rate between the known currency and desired currency.
-   *
-   * @param args - The arguments to this function.
-   * @param args.tokenAddresses - Addresses for tokens.
-   * @param args.chainId - The EIP-155 ID of the chain where the tokens live.
-   * @param args.nativeCurrency - The native currency in which to request
-   * exchange rates.
-   * @returns A map from token address to its exchange rate in the native
-   * currency, or an empty map if no exchange rates can be obtained for the
-   * chain ID.
-   */
-  async #fetchAndMapExchangeRates({
-    tokenAddresses,
-    chainId,
-    nativeCurrency,
-  }: {
-    tokenAddresses: Hex[];
-    chainId: Hex;
-    nativeCurrency: string;
-  }): Promise<ContractMarketData> {
-    if (!this.#tokenPricesService.validateChainIdSupported(chainId)) {
-      return tokenAddresses.reduce((obj, tokenAddress) => {
-        obj = {
-          ...obj,
-          [tokenAddress]: undefined,
-        };
-
-        return obj;
-      }, {});
-    }
+  async #fetchAndMapExchangeRatesForSupportedNativeCurrency(
+    assets: {
+      chainId: Hex;
+      tokenAddress: Hex;
+    }[],
+    currency: string,
+    marketData: Record<Hex, Record<Hex, MarketDataDetails>> = {},
+  ) {
+    return await reduceInBatchesSerially<
+      { chainId: Hex; tokenAddress: Hex },
+      Record<Hex, Record<Hex, MarketDataDetails>>
+    >({
+      values: assets,
+      batchSize: TOKEN_PRICES_BATCH_SIZE,
+      eachBatch: async (partialMarketData, assetsBatch) => {
+        const batchMarketData = await this.#tokenPricesService.fetchTokenPrices(
+          {
+            assets: assetsBatch,
+            currency,
+          },
+        );
 
-    if (this.#tokenPricesService.validateCurrencySupported(nativeCurrency)) {
-      return await this.#fetchAndMapExchangeRatesForSupportedNativeCurrency({
-        tokenAddresses,
-        chainId,
-        nativeCurrency,
-      });
-    }
+        for (const tokenPrice of batchMarketData) {
+          (partialMarketData[tokenPrice.chainId] ??= {})[
+            tokenPrice.tokenAddress
+          ] = tokenPrice;
+        }
 
-    return await this.#fetchAndMapExchangeRatesForUnsupportedNativeCurrency({
-      chainId,
-      tokenAddresses,
-      nativeCurrency,
+        return partialMarketData;
+      },
+      initialResult: marketData,
     });
   }
 
+  async #fetchAndMapExchangeRatesForUnsupportedNativeCurrency(
+    assets: {
+      chainId: Hex;
+      tokenAddress: Hex;
+    }[],
+    currency: string,
+    marketData: Record<Hex, Record<Hex, MarketDataDetails>>,
+  ) {
+    // Step -1: Then fetch all tracked tokens priced in USD
+    const marketDataInUSD =
+      await this.#fetchAndMapExchangeRatesForSupportedNativeCurrency(
+        assets,
+        'usd', // Fallback currency when the native currency is not supported
+      );
```

Comment:
This call recovers both tokens and native asset for the chain, so it saves us from doing two calls.

Reply:
niiiiiice!!